### DISCUSSION TITLE
Make it possible to provide types for tags

### DISCUSSION THREAD

**Discussion by JReinhold** [statelyai/xstate#2443] [Ideas]
The [tags feature](https://xstate.js.org/docs/guides/states.html#state-hastag-tag) is really nice and makes it easier to match for multiple states. However it's currently not possible to provide TypeScript typings for which tags are valid, which is a shame. I recently refactored some tags code and didn't change all parts, because TypeScript didn't warn me about invalid tag lookups and creatings.

From a user perspective I imagine something like this:

```typescript
type MyTags = 'firstTag' | 'secondTag';
const machine = createMachine<MyContext, MyEvent, MyState, MyTags>({
  ...
  states: {
    aState: {
      tags: 'firstTag' // valid
    },
    bState: {
      tags: 'otherTag' // TS error!
    }
});

...
state.hasTag('firstTag'); // valid
state.hasTag('doesNotExist'); // TS Error
```

I don't know how this would be implemented, but I would at least suggest that tags default to the `string` type so developers don't have to type it explicitly if they don't want to.

**Comment by Andarist** [comment]
Our list of generics is getting quite long. We want to provide the best-in-class TS experience but it might require taking some other approach that we need to figure out first.

This, among other things, is very much on our minds. It just won't happen overnight. Stay tuned though! This year will bring a lot of new exciting stuff. 🚀

**Reply by davidkpiano** [reply]
A future idea for this would be to _model_ the tags:

```js
const model = createModel({...}, {
  tags: ['error', 'pending', 'valid', 'invalid']
});

const machine = model.createMachine({
  // ...
  states: {
    loading: {
      tags: 'pending', // strongly typed
    }
  }
});
```
